File: src/core/loadTsv.ts
/**
 * Load a `key<TAB>value` TSV beside the calling module into a Map, skipping blank and `#`-comment lines.
 * Collapses the read + split + comment-skip + tab-split boilerplate that the per-language dictionary
 * loaders (stress / tone / rhyme / lexicon tables) otherwise repeat.
 *
 *   const STRESS = loadTsvMap(import.meta.url, "stress.tsv", Number, { optional: true });
 *
 * `parse(value, key)` maps the raw post-tab string to the stored value (default: the raw string). Returning
 * `undefined` skips the row (for loaders that filter, e.g. reject non-numeric values). `optional: true` makes a
 * missing file yield an empty Map instead of throwing (for lexicons that may be absent).
 */
import { dataFile } from "./dataPath.ts";
import { readDataText } from "./dataSource.ts";

/** Read a data file beside `metaUrl`, returning its non-blank, non-`#`-comment lines. `optional` → [] on a
 *  missing file (else rethrows). Shared by loadTsvMap and loadLines so both parse lines identically. */
function readDataLines(
    metaUrl: string,
    filename: string,
    optional: boolean,
): string[] {
    let text: string;
    try {
        text = readDataText(dataFile(metaUrl, filename));
    } catch (err) {
        if (optional) return [];
        throw err;
    }
    return text.split(/\r?\n/).filter((l) => l !== "" && !l.startsWith("#"));
}

/**
 * ⚠ `fold` MAKES A LEXICON REACHABLE THROUGH ITS OWN ENGINE'S NATIVISER (#1068). An engine folds a word to
 * its declared inventory BEFORE looking it up, so a headword spelled with a letter the fold rewrites can
 * never be matched from `text()` — it does not throw and it does not drop a phone, the word just takes the
 * OOV path and gets a plausible wrong reading. Passing the engine's own `nat` here adds the folded spelling
 * as an ALIAS for the same value, which is what closes the gap. Measured in
 * `docs/investigations/nativiser_lexicon_seam_investigation.md`; guarded by `test/lexicon-reachability`.
 *
 * ⚠ AN UNFOLDED KEY ALREADY IN THE FILE WINS, ALWAYS. An alias is written only into a FREE slot, so no
 * reading the engine already resolves THROUGH THE LEXICON can change. Without this, Slovene alone would
 * have 99 keys where the two spellings disagree about which nucleus is stressed (`bləste`=0 against
 * `bleste`=1) and the loser would be decided by file order.
 *
 * ⚠ …BUT THAT IS NOT "THE GOLDENS CANNOT MOVE", AND #1072 SAID SO AND WAS WRONG. A free slot is free
 * because the word was an OOV MISS, and an OOV miss is not silence — it is the engine's fallback rule,
 * which the goldens record. Slovene's 684 orphan aliases are 680 new headwords, and eight of `sl.tsv`'s
 * 200 rows moved from the penultimate fallback onto the lexicon (`umrl` → *umˈərl* where the fallback said
 * *ˈumərl*). No gate caught it because sl was unported at the time, so nothing ran its golden; the sl port
 * regenerated it. The invariant to state is about LEXICON-RESOLVED readings, not about goldens.
 *
 * ⚠ AND THE RAW KEY IS KEPT, not replaced. Aliasing is additive: some callers (evals, `phonemizeWord` used
 * bare) look words up without nativising first, and dropping the original spelling would break them for no
 * gain — the alias is what the shipped path needs, not the absence of the original.
 *
 * ⚠ ITERATION IS OVER THE FILE'S ROWS IN ORDER, not over the Map. Two different keys can fold onto the same
 * free slot, so "first one wins" must mean first IN THE FILE — a Map happens to preserve insertion order and
 * .NET's Dictionary does not, and a port that iterated the dictionary would silently pick a different winner.
 */
export function loadTsvMap<V = string>(
    metaUrl: string,
    filename: string,
    parse: (value: string, key: string) => V | undefined = (v) =>
        v as unknown as V,
    opts: { optional?: boolean; fold?: (key: string) => string } = {},
): Map<string, V> {
    const map = new Map<string, V>();
    const rows: Array<[string, V]> = [];
    for (const line of readDataLines(metaUrl, filename, opts.optional ?? false)) {
        const tab = line.indexOf("\t");
        if (tab <= 0) continue;
        const v = parse(line.slice(tab + 1), line.slice(0, tab));
        if (v !== undefined) {
            map.set(line.slice(0, tab), v);
            rows.push([line.slice(0, tab), v]);
        }
    }
    if (opts.fold) {
        for (const [k, v] of rows) {
            const f = opts.fold(k);
            if (f !== k && !map.has(f)) map.set(f, v);
        }
    }
    return map;
}

/**
 * Load a one-token-per-line membership list beside the calling module, skipping blank and `#`-comment lines.
 * The value-less counterpart of loadTsvMap, for word-lists that back a Set (Japanese adverbs, Thai seg-words).
 *
 *   const ADVERBS = new Set(loadLines(import.meta.url, "adverbs.txt"));
 *
 * `optional: true` makes a missing file yield [] instead of throwing.
 */
export function loadLines(
    metaUrl: string,
    filename: string,
    opts: { optional?: boolean } = {},
): string[] {
    return readDataLines(metaUrl, filename, opts.optional ?? false);
}

File: csharp/Vernacula.Phonemizer/Core/LoadTsv.cs
/**
 * Load a `key<TAB>value` TSV into a Map, and its value-less counterpart for membership word-lists;
 * blank and `#`-comment lines are skipped.
 * Ported from src/core/loadTsv.ts — see that file for the rationale.
 *
 * C# PORT NOTE: the TS resolves data beside the calling module via `import.meta.url`. Here the caller
 * passes `moduleDir`, its directory RELATIVE TO the repo's `src/` (e.g. "core", "languages/amharic"),
 * and the file is resolved through DataPath — the data files are not ported.
 */
namespace Vernacula.Phonemizer.Core;

public static class LoadTsv
{
    private static readonly JsRe LineSplit = JsRegex.Compile("\\r?\\n");

    /** Non-blank, non-`#`-comment lines of a data file; `optional` → [] on a missing file. Shared by
     *  LoadTsvMap and LoadLines so both parse lines identically. */
    private static List<string> ReadDataLines(string moduleDir, string filename, bool optional)
    {
        // Optionality is decided HERE, not in DataPath, which throws on a missing file by design.
        var path = DataPath.ResolveAllowMissing(moduleDir + "/" + filename);
        string text;
        try
        {
            text = File.ReadAllText(path);
        }
        catch (IOException) when (optional)
        {
            return new List<string>();
        }
        catch (SystemException e) when (optional && e is FileNotFoundException or DirectoryNotFoundException or UnauthorizedAccessException)
        {
            return new List<string>();
        }
        var lines = new List<string>();
        foreach (var l in LineSplit.Re.Split(text))
            if (l != "" && !l.StartsWith("#", StringComparison.Ordinal))
                lines.Add(l);
        return lines;
    }

    /// <summary>Insertion-ordered like a JS Map (C# Dictionary preserves insertion order as long as nothing
    /// is removed, which holds here — noted because some consumers iterate).</summary>
    public static Dictionary<string, string> LoadTsvMap(string moduleDir, string filename, bool optional = false) =>
        LoadTsvMap<string>(moduleDir, filename, (v, _) => v, optional);

    /// <summary>⚠ <c>fold</c> MAKES A LEXICON REACHABLE THROUGH ITS OWN ENGINE'S NATIVISER (#1068). An engine
    /// folds a word to its declared inventory BEFORE looking it up, so a headword spelled with a letter the
    /// fold rewrites can never be matched from Text() — the word silently takes the OOV path and gets a
    /// plausible wrong reading. Passing the engine's own nativiser adds the folded spelling as an ALIAS for
    /// the same value.
    /// <para>⚠ AN UNFOLDED KEY ALREADY IN THE FILE WINS, ALWAYS — an alias is written only into a FREE slot,
    /// so no LEXICON-RESOLVED reading can change. ⚠ That is not "no golden moves": a free slot is free
    /// because the word was an OOV miss, and the fallback rule's answer is in the goldens. See the TS.</para>
    /// <para>⚠ AND THE ITERATION IS OVER THE FILE'S ROWS IN ORDER, which is why the rows are collected into a
    /// list first. Two keys can fold onto the same free slot, and "first wins" only means anything if the
    /// order is the file's. A Dictionary happens to preserve insertion order here, but relying on that for a
    /// tie-break would be relying on an implementation detail the TS side does not share.</para>
    /// See docs/investigations/normalization/nativiser_lexicon_seam_investigation.md and test/lexicon-reachability.</summary>
    public static Dictionary<string, V> LoadTsvMap<V>(
        string moduleDir,
        string filename,
        Func<string, string, V?> parse,
        bool optional = false,
        Func<string, string>? fold = null) where V : class
    {
        var map = new Dictionary<string, V>();
        var rows = new List<(string Key, V Value)>();
        foreach (var line in ReadDataLines(moduleDir, filename, optional))
        {
            var tab = line.IndexOf('\t');
            if (tab <= 0) continue;
            var v = parse(line[(tab + 1)..], line[..tab]);
            if (v is not null) { map[line[..tab]] = v; rows.Add((line[..tab], v)); }
        }
        if (fold is not null)
            foreach (var (k, v) in rows)
            {
                var f = fold(k);
                if (f != k && !map.ContainsKey(f)) map[f] = v;
            }
        return map;
    }

    /// <summary>Value-type variant of <see cref="LoadTsvMap{V}(string,string,Func{string,string,V},bool)"/>:
    /// the TS `Number` parses become `double?` here, and a null skips the row as `undefined` does there.
    /// <para>⚠ <c>fold</c> CARRIES THE SAME CONTRACT AS THE REFERENCE-TYPE OVERLOAD — unfolded key wins, and
    /// the alias pass iterates the FILE'S ROWS IN ORDER rather than the dictionary. See that overload.</para></summary>
    public static Dictionary<string, V> LoadTsvMapV<V>(
        string moduleDir,
        string filename,
        Func<string, string, V?> parse,
        bool optional = false,
        Func<string, string>? fold = null) where V : struct
    {
        var map = new Dictionary<string, V>();
        var rows = new List<(string Key, V Value)>();
        foreach (var line in ReadDataLines(moduleDir, filename, optional))
        {
            var tab = line.IndexOf('\t');
            if (tab <= 0) continue;
            var v = parse(line[(tab + 1)..], line[..tab]);
            if (v.HasValue) { map[line[..tab]] = v.Value; rows.Add((line[..tab], v.Value)); }
        }
        if (fold is not null)
            foreach (var (k, v) in rows)
            {
                var f = fold(k);
                if (f != k && !map.ContainsKey(f)) map[f] = v;
            }
        return map;
    }

    /** One token per line, for word-lists that back a HashSet. */
    public static List<string> LoadLines(string moduleDir, string filename, bool optional = false) =>
        ReadDataLines(moduleDir, filename, optional);
}
